pkg/scheduler/framework/plugins/defaultpreemption/default_preemption.go

```diff
@@ -53,21 +56,95 @@ const (
 // DefaultPreemption is a PostFilter plugin implements the preemption logic.
 type DefaultPreemption struct {
 	fh        framework.FrameworkHandle
+	args      config.DefaultPreemptionArgs
 	podLister corelisters.PodLister
 	pdbLister policylisters.PodDisruptionBudgetLister
+	// candidateOffset maintains state about the last offset used for dry running
+	// preemption.
+	candidateOffset int32
 }
 
 var _ framework.PostFilterPlugin = &DefaultPreemption{}
 
+type FindCandidatesOptions struct {
+	CandidateOffset             *int32
+	MinCandidateNodesPercentage int32
+	MinCandidateNodesAbsolute   int32
+}
+
+type FindCandidatesOption func(*FindCandidatesOptions)
+
+// calculateNumCandidates returns the number of candidates the FindCandidates
+// method must produce from dry running based on the constraints given by
+// <minCandidateNodesPercentage> and <minCandidateNodesAbsolute>. The number of candidates
+// returned will never be greater than <numNodes>.
+func (fco *FindCandidatesOptions) calculateNumCandidates(numNodes int32) int32 {
```

Comment:
this needs to be calculated once, lets just do that at New and store it in a plugin variable.

Reply:
So that we only need 2 options: WithCandidateOffset() and WithNumCandidates().

Reply:
In that case, out-of-tree that has its own pluginArgs (or empty pluginArgs) won't be able to get the benefits.

Reply:
yeah, the idea is to make `FindCandidates` stateless

Reply:
I think we are lacking proper separation between what we want to export for out-of-tree preemption plugins and what is specific to the default preemption plugin. Having everything here is honestly bad and confusing.

Reply:
Agree. We should polish the logic, maybe in a follow-up. Also I'm not quite satisfied with the current function signature - to make it stateless, it even has 8 parameters! Which is way too much, needs to shrink the args number a bit.

Reply:
https://github.com/kubernetes/kubernetes/issues/95131

Reply:
I still don't think we need all of that, I expect the refactoring to result in a new type (e.g. `PreemptionEvaluator`) and a group of member methods. The refactoring will basically move the state to this new type along with the functions.